Lineage UI documentation (#544)
astronomer/docs: astro/data-lineage.md

---
sidebar_label: 'Data Lineage'
title: "Data Lineage on Astro"
id: data-lineage
description: "Track and visualize the movement of your data with data lineage on Astro"
---

## Overview

This guide explains how to navigate the **Lineage** page in the Cloud UI and configure views that can help you both troubleshoot issues with your data pipelines as well as understand the movement of data across your Organization.

> Data lineage is a critical part of data orchestration tooling.  The **Lineage** tab on Astro has three pages:

Author: "Astro homepage" isn't a term we've used before, why wouldn't we say Cloud UI here?

> - **Explore**: A real-time overview of all jobs that emit data lineage across your Organization

Author: We can only anchor on job if it's a header in the doc, for now I can include a general link to the concepts doc at the end of the overview section.

> - **Explore**: A real-time overview of all jobs that emit data lineage across your Organization

Reviewer: Do you think it's overkill then to perhaps just add another sentence at the end here saying: `In Airflow a Task maps to an OpenLineage Job` ?

> - **Explore**: A real-time overview of all jobs that emit data lineage across your Organization

Reviewer: I just assume that reading this page could be a lot of users first exposure to the term "job".

> - **Explore**: A real-time overview of all jobs that emit data lineage across your Organization

Author: I added "A job can be an Airflow task or another process configured to emit lineage data to Astronomer, such as a Spark job."

- **Issues**: A view of potential issues or statistical inconsistencies related to your jobs or datasets
- **Lineage**: A graph view that visualizes data movements across DAGs, tasks, and Deployments

Generally speaking, you can use these pages to diagnose issues that may otherwise be difficult to identify across environments and tools. For example, if an Airflow task failed because the schema of a database changed, you might go to the **Lineage** page on Astro to determine which job caused that change and which downstream tasks failed because of it.

Views in the **Lineage** page are available to all members of your Organization on Astro.

> ## View the Lineage Graph for a Deployment

Author: The lineage graph includes all of these things, so I'm not sure it makes sense to say the "graph of a [component]". A lineage graph is organized by Deployment/namespace, right @conorbev?

> To view the lineage graph for one of your Deployments, search for a DAG or task in the Deployment with the search bar at the top of the UI.

Author: I would add "from other tools with lineage integrations"

![Example query in the lineage search bar](/img/docs/lineage-search.png)

> Clicking a search result will open the **Lineage** page and show the lineage graph for the Deployment's most recent DAG run.

Author: Might make sense to include that other note in either the compatibility or integrations doc, since it's a little more actionable there. Feels slightly out of place here.

You can also access the lineage graph for a recent DAG run in the **Explore** page under **Most Recent Runs**.

:::info

> By default, when you directly access the **Lineage** page from the left-hand menu, the page shows your most recently accessed lineage graph. If you go to the **Lineage** page before first accessing a lineage graph, you will not see any lineage graph data. If this happens, try first accessing a task using the search bar or **Explore** page to populate the **Lineage** page with data.

Author: Modified this slightly

:::

:::info

The **Lineage** page shows lineage data only for the most recent run of a given Deployment. To explore lineage data from previous runs, see [Compare Lineage Graphs from Previous Runs](data-lineage.md#compare-lineage-graphs-from-previous-runs).

:::

## Navigating the Lineage Graph

In the **Lineage** page, Astronomer renders your data pipeline as a directed graph of **job** and **dataset** nodes:

- A **job** node represents an individual step in your data pipeline, such as an Airflow task in your Deployment or a Spark job.
- A **dataset** node represents a data source that your tasks interact with, such as a Snowflake database.

Directed vertices connect jobs to datasets and vice versa. A single vertex will never connect two jobs or two datasets together.

> In the following example, `insert` is a job that exists as part of the `etl_menu_items` grouping. A vertex connects `insert` to the `menu_items` dataset to indicate that `insert` interacted with data in this dataset.

Author: I don't mind `group`

## View Metrics for a Specific Job or Dataset

Below the lineage graph is the **information pane**: a collection of information and metrics for a single selected node.

The information pane is split into the following tabs:

- **Info**: Shows the code for a job or the schema for a dataset. Also shows the difference between job runs when you create a comparison in the **Compare** tab
- **Inputs/Outputs**: Shows the inputs and outputs for a job or dataset. This information is equivalent to the upstream and downstream nodes in the graph view
- **Quality (Dataset only)**: Shows the data quality checks performed on each element of a dataset. You can drill down further into these checks by expanding a listed dataset element
- **Duration (Job only)**: Shows the duration of upstream job runs, starting with the most upstream job run and descending to the currently selected job run
- **Compare (Job only)**: Shows other job runs of the currently selected job. Select any two job runs and go to the **Info** tab to see how the code changed between the two job runs. Use this tab to compare job runs with different statuses or run times to measure performance between code changes

> ### Using the Quality Tab

Reviewer 2: ```suggestion
### Check for Data Quality
```

> The **Quality** tab appears in the information pane whenever you select a dataset node from the graph. It contains both high level charts that provide information about the entire dataset, as well as charts for each individual column of the dataset. These graphs appear only if there is underlying metadata generated by an integrated data check tool, such as Great Expectations.

Reviewer 2: ```suggestion
Whenever you select a dataset from the lineage graph, the **Quality** tab appears in the information pane. This tab shows a list of columns that correspond to your dataset.

If you integrate with [Great Expectations](https://docs.greatexpectations.io/docs/), this tab also shows:

- The number of rows in your dataset over time
- The amount of megabytes consumed by your dataset over time
- Successes and failures of data quality checks

If Great Expectations is not configured, this metadata will not be available.
```

![Quality tab example](/img/docs/quality-tab.png)

> Use the **Quality** tab to detect looking for sudden, unexplained changes in your dataset that could indicate an upstream failure. The following topics explain each available chart in the tab.

Reviewer 2: ```suggestion
Use the **Quality** tab to detect unexpected changes or statistical variance in your dataset that could indicate a problem. The following topics explain each available chart in the tab.
```

#### Rows

The Rows chart shows the total number of rows in the dataset over time. A drastic change in rows can occur naturally (for example: when a rapid increase of customer orders during the holiday season). However, it can also indicate an error in an upstream job, especially if it is sudden or unexpected.

#### Bytes

> The Bytes chart shows the total size of the dataset over time. A sudden increase in dataset size usually means something has changed in the definition of the data. For example, a new column might have been added to your table containing the description of an order, where before it contained only part numbers and quantities.

Author: Holding off on this info for now based on Slack discussion

    ![Info tab when comparing two code sources](/img/docs/compare-code.png)

## View a Summary of Issues across Deployments

The **Issues** page contains metrics that can help you identify irregularities related to the behavior of your data pipelines.

![Lineage issues page](/img/docs/lineage-issues.png)

Specifically, this page tracks following types of issues

- **Job Execution Issues**: A job execution issue occurs when a job emits an error that it did not successfully complete. Note that this metric works only for Deployments using Runtime 5.0.0+.
- **Job Duration Issues**: A job duration issue occurs when a job run's duration is more than three standard deviations longer than the average run time for that specific job.
- **Data Quality Issues**: If you integrate with [Great Expectations](https://www.astronomer.io/guides/airflow-great-expectations/), an open source data quality tool, this metric will track issues generated by expectation quality checks for both your jobs and datasets. Use this tab to detect data quality failures that could indicate an upstream problem.

## View a Summary of Past Runs

> By default, the **Lineage** page shows the last lineage graph you accessed. To see high level metrics about all of your job runs and access lineage graphs for different Deployments, open the **Explore** page on the lefthand sidebar. This page is structured similarly to the Airflow UI's calendar view: It contains a list of your most recent runs, as well as a calendar that shows all runs over the last year.

Reviewer: remove: `for different Deployments`

I think we should add something like:

`For Airflow tasks the job name will be <dag_id>.<task_id>. For example the job: "example_dag_basic.extract" indicates the **extract** task running within the **example_dag_basic** DAG.`

This guide explains how to navigate the **Lineage** tab in the Cloud UI and configure views that can help you troubleshoot issues with your data pipelines and understand the movement of data across your Organization.

The **Lineage** tab on Astro has three pages:

- **Explore**: A real-time overview of all **jobs** that emit data lineage across your Organization. A job can be an Airflow task or any other process configured to emit lineage data to Astronomer, such as a Spark job.
- **Issues**: A view of potential issues or statistical inconsistencies related to your jobs or datasets
- **Lineage**: A graph view that visualizes data lineage

Generally speaking, you can use these pages to diagnose issues that may otherwise be difficult to identify across environments and tools. For example, if an Airflow task failed because the schema of a database changed, you might go to the **Lineage** page on Astro to determine which job caused that change and which downstream tasks failed because of it.

Views in the **Lineage** page are available to all members of your Organization on Astro. For more information on data lineage and related concepts, see [Lineage Overview](lineage-overview.md).

> 

Author: ```suggestion

:::info 
This functionality is early access and under active development. If you have any questions or feedback about this feature, reach out to [Astronomer Support](https://support.astronomer.io/).
:::

```

## View the Lineage Graph for a Data Pipeline

To view the lineage graph for one of your data pipelines, search for a DAG, task, or dataset with the search bar at the top of the UI. You can also search for jobs from other tools with lineage integrations, including dbt or Spark.

![Example query in the lineage search bar](/img/docs/lineage-search.png)

> The search results include the namespace that emitted the matching event. If an Astro Deployment emitted the lineage event, then this namespace will match Deployment's namespace as shown in the **Deployments** page of the Cloud UI. Clicking a search result will open the **Lineage** page and show the lineage graph for the Deployment's most recent job run.

Reviewer: I think: `show the lineage graph for the Deployment's most recent job run.` --> `show the lineage graph for the selected job or dataset.`

You can also access the lineage graph for a recent job run in the **Explore** page under **Most Recent Runs**.

> 

Reviewer: Reading this again, I think we should add another info section something like:

```
:::info

If the Lineage graph only shows you a single node it indicates that the job you selected did not emit any input or output datasets, likely because it is from a Task isn't using an [Airflow Operator that is supported out of the box.](link to supported operators doc). You will still be able to view the duration of this job over time. 
```

^ if we think 3 info boxes is too many I feel like this one ( or some edited version of it ) is probably even more important than the "most recent run" one.

To learn more information about a dataset or job, you can either hover over or click a node. Hovering over a node gives you high level information about the dataset or job at a glance. Specifically, you'll see:

- **Namespace**: The namespace of the Deployment in which the job ran
- **Name**: The DAG ID and task ID of the job, formatted as `<dag-id>.<task-id>`
- **Run information (job only)**: Metadata and status information about the job run
- **Quality checks (dataset only)**: The status of a dataset's data quality checks

Clicking a node populates the information pane with detailed information about the job or dataset. For more information about how to use this view, see [View Metrics for a Specific Job or Dataset](data-lineage.md#view-metrics-for-a-specific-job-or-dataset).

### Access the Graph Legend

Clicking on the key icon in the information pane opens the graph legend. The legend provides a visual guide to help you distinguish between:

- Job nodes and dataset nodes.
- User-selected and unselected nodes.
- Completed, running, failed, and aborted jobs
- Completed, running, failed, and aborted dataset checks

:::info

> Dataset and job statuses are based on checks of metadata attached to your data. To see the specifics of why a database check failed, you can click on the failed database node and check the **Quality** tab in the information pane. To check why a job failed, you might need to check the source of the job, such as your DAG or task.

Reviewer: `database check failed` --> `dataset failed`
